이모지를 채널에 날릴 수 있다!🤭v2

happydhKim opened the pull request:
# [Epic#38] 이모지를 채널에 날릴 수 있다!🤭v2

## 해당 이슈 📎

- [Epic#38] 이모지를 채널에 날릴 수 있다!🤭v2 (#159)

## 변경 사항 🛠

- 이제 날아가는 이모지가 정상적인 위치에서 출발합니다.
- 날아가는 이모지가 작아졌어요..😥
- 날아가는 이모지가 너무 많을경우 출발을 거절합니다.🤭

## 테스트 ✨

>없음

## 리뷰어 참고 사항 🙋‍♀️

> 없음

happydhKim committed:
[ UTIL ] 날아가는 이모지 절대좌표 UTIL 구현 (#87)

Why
- 날아가는 이모지의 절대좌표가 필요하다.
- 기존 클라이언트 클릭 이벤트 좌표는 각 클라이언트 화면 별 다르게 출력되는 버그가 있다.

What
- 날아가는 이모지의 절대좌표구하는 함수 구현.
- 총 3개의 element 좌표를 각 이벤트 발생별 구하는 함수를 구현.

happydhKim committed:
[ FEATURE ] 날아가는 이모지 좌표 스키마 수정 및 동기화 (#159)

Why
- 날아가는 이모지의 스키마에 좌표가 필요 없어졌다. (UTIL생성)

What
- 날아가는 이모지 스키마 수정. ( 좌표 관련 삭제 )
- 날아가는 이모지 graphQL과 연동.

happydhKim committed:
[ FEATURE ] 날아가는 이모지 크기 축소 및 풀스크린 좌표 수정 (#159)

Why
- 날아가는 이모지의 크기가 조금 크다는 의견이 있어서 수정하기로 결정.
- 전체화면의 좌표값을 구하는게 일정하다.(변화가 되지 않는다.)

What
- 날아가는 이모지의 크기를 축소.
- 전체화면의 좌표값을 구하는 변수를 함수로 바꿈으로 써, 매번 좌표의 새로운 값을 구한다.

happydhKim committed:
[ FEATURE ] 날아가는 이모지의 갯수 설정 (#159)

Why
- requestAnimationFrame의 특성상 다른 화면일 시 애니메이션을 쌓아둔다.
- 그렇기 때문에 한번에 수 많은 이모지를 대기시켰다가 한번에 작동 시키면 브라우저가 터질 수도 있다.

What
- 최대 제한을 설정하는 PREVENT_FLYING_EMOJI 함수를 구현.
- 최대 날아가는 이모지를 200개로 설정하여, 그 이상일 경우 대기시키지 못하게 한다.

happydhKim committed:
[ REFACTOR ] 날아가는 이모지 스크롤을 고려한 좌표 설정 (#159)

Why
- 스크롤을 움직였을 경우 이모지가 이상한 곳에서 날아간다.

What
- 스크롤을 계산해서 좌표를 정확하게 설정하도록 구현.

happydhKim committed:
[ FIX ] 날아가는 이모지의 최대 갯수 수정 (#159)

Why
- 날아가는 이모지의 숫자가 약간 적은듯 하다.

What
- 200개에서 300개로 수정.

happydhKim committed:
[ FIX ] 날아가는 이모지 전체화면 절대좌표 수정 (#159)

Why
- 전체화면에서 body로 설정 했을 경우 이상한 곳에서 이모지가 날아간다.

What
- body가 아닌 fullscreen에서 절대좌표를 구해 출발하도록 수정.

happydhKim committed:
[ REFACTOR ] 네이밍이 잘못된 변수명 수정 (#159)

Why
- 네이밍이 잘못된 변수명을 수정해야 한다.

What
- 잘못된 변수명 수정.

happydhKim committed:
[ FIX ] 날아가는 이모지 onfocus 상태에만 나오도록 수정 (#159)

Why
- 날아가는 이모지가 onfocus인 상태에만 나오도록 수정할 필요가 있다.
- onblur 상태에서 requestAnimationFrame이 이모지를 쌓아 두기 때문이다.

What
- onfocus와 onblur로 관리하는 상태를 추가 했다.

happydhKim committed:
[ MERGE ] 최신 버전으로 업데이트

Why
- 최신버전의 develop으로 업데이트 해야 한다.

What
- 최신버전의 코드로 업데이트

devjaewon commented on web/src/constants/flyingEmoji.js:
```diff
@@ -1,15 +1,48 @@
 export const FULL_SCREEN_SCALE = () => 4.5 + Math.random();
-export const NORMAL_SCREEN_SCALE = () => 3 + Math.random();
+export const NORMAL_SCREEN_SCALE = () => 2.5 + Math.random();
 export const OPACITY = 1.0;
```
Computed Constants 들도 이렇게 상수형태로 빼놓 은것 좋네요!!

devjaewon reviewed (approved):
고생하셨습니다!! 이모지 기능 완성도가 점점 올라가고 있어서 좋네요 👍🏻

happydhKim committed:
[ FIX ] 이모지의 최대 갯수 1000개로 설정 (#159)

Why
- 많은 사람들이 동시에 요청을 했을 경우를 대비할 필요가 있다.

What
- 최대 1000개의 날아가는 이모지가 있다면, 더이상 못날아가도록 설정.
